docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,303 @@
+- Feature Name: table_descriptor_lease
+- Status: draft
+- Start Date: 2015-10-09
+- RFC PR:
+- Cockroach Issue: [#2036](https://github.com/cockroachdb/cockroach/issues/2036)
+
+# Summary
+
+Implement a table descriptor lease service to allow safe usage of
+cached table descriptors.
+
+# Motivation
+
+Table descriptors contain the schema for a single table and are
+utilized by nearly every SQL operation. Fast access to the table
+descriptor is critical for good performance. Reading the table
+descriptor from the KV store on every operation adds significant
+latency.
+
+Table descriptors are currently distributed to every node in the
+cluster via gossipping of the system config (see
+[schema_distribution](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/schema_gossip.md)). Unfortunately,
+it is not safe to use these gossipped table descriptors in almost any
+circumstance. Consider the statements:
+
+```sql
+  CREATE TABLE test (key INT PRIMARY KEY, value INT);
+  CREATE INDEX foo ON test (value);
+  INSERT INTO test VALUES (1, 2);
+```
+
+Depending on when the gossip of the schema change is received, the
+`INSERT` statement may either see no cached descriptor for the `test`
+table (which is safe because it will fall back to reading from the KV
+store), the descriptor as it looks after the `CREATE TABLE` statement
+or the descriptor as it looks after the `CREATE INDEX` statement. If
+the `INSERT` statement does not see the `CREATE INDEX` statement then
+it will merrily insert the values into `test` without creating the
+entry for the `foo` index, thus leaving the table data in an
+inconsistent state.
+
+It is easy to show similar problematic sequences of operations for
+`DELETE`, `UPDATE` and `SELECT`. Fortunately, a solution exists in the
+literature: break up the schema change into discrete steps for which
+any two consecutive versions of the schema are safe for concurrent
+use. For example, to add an index to a table we would perform the
+following steps:
+
+* Add the index to the table descriptor but mark it is as
```

Comment:
I believe the state transition schema here is 

S1 -> S2-delete-only -> S2-write-only -> S2

in the delete-only state , delete/update operations are allowed to delete index values for rows that are deleted/updated.